### ISSUE TITLE
Is BERT fine-tuned for B-AON?(Discrepency in results)

### ISSUE THREAD

**Issue by RuskinManku** [shrimai/Topological-Sort-for-Sentence-Ordering#4]
In the paper, you also report results for B-AON, which simply replaces the LSTM of AON with the pre-trained BERT model as Sentence Encoder. Do you directly extract the BERT embeddings for sentences and then train the model, or is BERT fine-tuned during the training?

**Comment by RuskinManku** [comment]
Okay, I just read the part where you mention you fine-tune B-AON. I have a follow-up question.
I'm not sure if you are still keeping up with the problem of Sentence Ordering, but recently the authors of AON also published another paper in EMNLP 2020, "[BERT-enhanced Relational Sentence Ordering Network](https://www.aclweb.org/anthology/2020.emnlp-main.511/)".

In that, they report results for a model they call, "BertSenPD", which essentially replaces the ranking network of RankTxNet ([Deep Attentive Ranking Networks for Learning to Order Sentences](https://arxiv.org/abs/2001.00056)) with a pointer network. Now, as per my understanding, this is the same as B-AON. BERT as sentence encoder( that's how RankTxNet do it), transformer layers to encode paragraph, and then a pointer network on the average paragraph embedding. However, the results reported for BertSenPD vary significantly from B-AON( for example, for SIND, PMR of BertSenPD is 19.19, whereas for B-AON it's 14.30). Am I wrong that BertSenPD and B-AON are the same, or is there some discrepancy here

**Comment by shrimai** [comment]
I quickly skimmed through the papers you mentioned. It looks like there a few differences: (1) B-AON uses a LSTM decoder because AON uses LSTM decoder. The BertSendPD and RankTxNet use a transformer decoder. (2) The loss objectives of RankTxNet and hence BertSenPD seem to be different than B-AON/AON. Finally, I had used the implementation of the AON as provided by the authors of the original paper and replaced BERT sentence embeddings. You must confirm if the BertSenPD implementation is similar to AON or RanTxNet (could be something small like number of layers, dimension size, beam size etc).

**Comment by RuskinManku** [comment]
First of all, thank you for taking the time out to reply and skimming through the papers.
1) RankTxNet uses a transformer as an encoder only. It is titled "Paragraph Encoder" in the paper, and they basically use FFNN layers on top of the transformer output. Also, in BertSenPD, the authors mention they replace the ranking module with "traditional pointer network decoder", which I am sure is the same as the decoder used in AON, as there is no mention of transformer decoder.
2) True, the loss objective for RankTxNet is different, but I guess that's what the authors of BertSenPD mean when they replace the "Ranking module" and used a traditional pointer decoder. Also, using BertSenPD as a baseline really makes sense for their final model BERSON.
3) Finally, I coded the implementation of B-AON all by myself after posting the issue here, although I only tested it on SIND, the results I got match the ones you reported. So yes I think I am misinterpreting what BertSenPD is, but from the details they have given, it is the same as B-AON is the best conclusion I can draw. 
4) I have matched some of the hyperparameters that they have reported, but the difference is still substantial.

In any case, I tried to contact the authors of BertSenPD's paper but they haven't responded, yet.

**Comment by shrimai** [comment]
Yes, I suppose only the authors of "BERT-enhanced Relational Sentence Ordering Network" can clarify what it means implementation-wise to replace ranking module with "traditional pointer network decoder" (in terms of implementation of the decoder and the loss used). I am glad that you were able to reproduce the results of B-AON on SIND. Note that the AON results reported in our work are also slightly different than the original paper (the reported results are our runs of the code provided by the authors). Best of luck with your experiments! Please close the issue if all your concerns are addressed.

**Comment by RuskinManku** [comment]
Right, let's see if they reply but I'm not hopeful. I will just implement their final model BERSON and see if the results match. Thanks for your time!
